src/tool/xmeta/cpp/models/struct_model.h

```diff
@@ -32,72 +38,11 @@ namespace xlang::xmeta
             m_fields.emplace_back(std::move(field));
         }
 
-        void resolve(std::map<std::string, class_type_semantics> const& symbols, xlang_error_manager & error_manager)
-        {
-            for (auto & field : m_fields)
-            {
-                type_ref & field_type = field.first;
-                if (!field_type.get_semantic().is_resolved())
-                {
-                    // TODO: Once we have using directives, we will need to go through many fully_qualified_ids here
-                    std::string ref_name = field_type.get_semantic().get_ref_name();
-                    std::string symbol = ref_name.find(".") != std::string::npos 
-                        ? ref_name : this->get_containing_namespace_body()->get_containing_namespace()->get_fully_qualified_id() + "." + ref_name;
+        void resolve(symbol_table & symbols, xlang_error_manager & error_manager);
 
-                    auto iter = symbols.find(symbol);
-                    if (iter == symbols.end())
-                    {
-                        error_manager.write_unresolved_type_error(get_decl_line(), symbol);
-                    }
-                    else
-                    {
-                        field_type.set_semantic(iter->second);
-                    }
-                }
-            }
-        }
+        bool has_circular_struct_declarations(xlang_error_manager & error_manager);
```

Comment:
this feels like a semantic check that shouldn't be baked directly into the model

Reply:
I've baked alot of the semantic checking into the models. Especially to resolve type refs within it.

Reply:
I think by making the model semantically check itself. I can easily use my walker to go through the models and call the semantic checking methods. The only downside is that the  models become bigger with these extra methods.

Reply:
type ref resolution should be baked in, but I worry that baking in other types of validation will make it harder to add semantic checks down the road.

Reply:
I will have that in mind when I write validation stuff.